### PR TITLE
`Quiz exercises`: Add visual editor for multiple choice questions

### PR DESCRIPTION
**Pull request by sebastiannikles**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
When creating/editing multiple-choice quiz questions, the only option to specify all the information including question, hints, explanations and answer options is a text editor, which uses markup.

While this should not be too much of a problem for users that are used to writing markup, this can be quickly overwhelming and confusing for inexperienced users as a study for my Master's thesis has shown.

A visually simpler approach could help such users here.

### Description
Other LMS such as Canvas offer a more visual mode when it comes to editing and creating quiz questions. Here, the question is shown more or less like it will be in the final rendering and the user can just type in all the information.

I've taken that approach for Artemis as well and users can now switch to the visual mode for multiple choice questions where the question and the different answer options are shown more clearly and not just in text form. It's possible to switch between the modes as the markdown will be parsed again when switching.

For now, this editor is very simple and supports only text input. In the future, it could be worthwhile supporting more input options such as files or styling options here.

### Steps for Testing

Prerequisites:
- 1 Instructor who can create quizzes
- 1 course

1. Log in to Artemis
2. Navigate to Course Management
3. Navigate to the exercises page
4. Create a new quiz
5. Add a multiple choice question
6. Switch to the visual mode to edit the question

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

Bildschirmfoto 2022-11-28 um 23 58 41
Bildschirmfoto 2022-11-28 um 23 58 55

### Test Coverage
| Class/File | Stmt | Branch | Funcs | Line |
|------------|-------:|-----:|-------:|-----:|
| multiple-choice-visual-question.component.ts  |     100 |    83.33 |     100 |     100 |

### CODE REVIEW

**Review by LiaLism — DISMISSED** [review]
Visual Editor is easy to use and works as expected. Tested on ts1.

**Review by Yunqing-Wang — APPROVED** [review]
The function works well, tested on ts 1. But as I tried to submit the quiz, no result will be shown after 1min although I refreshed the website
image

**Comment by krusche** [comment]
please find more (experienced) reviewers

**Review by timo07k — DISMISSED** [review]
Visual editor works well. Tested on TS1.

**Review by evgeny1984 — CHANGES_REQUESTED** [review]
The feature works good, tested on ts1. 2 suggestions:

1. Distinction between wrong and correct answer using icon or color
2. I can only create the correct answer option in visual editor, not the wrong one.

**Review by PhilipWerz — DISMISSED** [review]
Visiual editor works good. Have tried a few things. Tested on TS1.

Note: 
- a bit weird that one has to remove standard text from answer field (but is the same for normal edit mode)

**Review by Santia-go — COMMENTED** [review]
Agreed with @evgeny1984 
All options by default are green, this could be confusing for the creators.  
Keep the wrong answer in other color (red preferible)

**Review by tunargs — APPROVED** [review]
Tested on ts1. Works as intended

**Review by rriyaldhi — CHANGES_REQUESTED** [review]
Functionality looks good, but left some suggestions

File: src/main/webapp/app/exercises/quiz/shared/questions/multiple-choice-question/multiple-choice-visual-question.component.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    styleUrls: ['./multiple-choice-question.component.scss', '../../../participate/quiz-participation.scss'],
+    encapsulation: ViewEncapsulation.None,
+})
+export class MultipleChoiceVisualQuestionComponent {
+    newOption = new AnswerOption();
+
+    _question: MultipleChoiceQuestion;
+
+    @Input()
+    set question(question: MultipleChoiceQuestion) {
```
[the remark this answers is not in the document]
Comment (sebastiannikles) [reply src/main/webapp/app/exercises/quiz/shared/questions/multiple-choice-question/multiple-choice-visual-question.component.ts:19]: We need a two-way binding here so the parent component noticed when the question was changed in the visual mode and can update the views (the warning for instance) accordingly.

**Review by rriyaldhi — APPROVED** [review]
Code looks good to me. Thanks for applying the suggestions.

**Review by frederik-kt — COMMENTED** [review]
There is some padding missing in the question section. This is Chrome.
Screenshot (94)

**Review by manuelmanso — DISMISSED** [review]
Tested on ts2, works as expected. 

I have a suggestion for the behavior of adding answer options. Instead of always showing the "New Answer Option" section and having the user fill in the fields and then click the "+" to actually add the option. Instead, I suggest only showing a "+" button to add a new option, and once the button is clicked the new option is alreadyv valid. For me this is preferable since in the current way it's possible for a user to fill in a new answer option but forget to click the "+" button.

**Review by danielbode — DISMISSED** [review]
Tested on ts2. The visual mode works fine.

I have one improvement suggestion. There is this toolbar in the edit mode and if you switch to the visual mode it disappears:

Bildschirm­foto 2023-02-08 um 18 07 41

I think this toolbar is especially useful in the visual tab because like you said in the PR description it mostly targets users that aren't familiar with markdown and they otherwise might not know how to make text bold or italic.

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested on TS2 in the testing session, looks nice but a few improvements to the visuals just like stated by the others would be nice

File: src/main/webapp/content/scss/themes/_dark-variables.scss [hunk 2]
```diff
@@ -337,6 +337,13 @@ $quiz-scoring-info-student-modal-question-wrong: $warning;
 $quiz-question-background: $neutral-dark;
 $quiz-question-border-color: $neutral-dark-l-20;
 
+$quiz-visual-question-border-color: #d3d3d3;
+$quiz-visual-answer-correct-container-color: #53a351;
+$quiz-visual-answer-correct-container-hover-color: #90c48e;
+$quiz-visual-answer-wrong-container-hover-color: #f3837a;
+$quiz-visual-answer-add-container-color: #508ac7;
+$quiz-visual-answer-add-container-hover-color: lighten($quiz-visual-answer-add-container-color, 20%);
+
```
Comment (bassner) [inline src/main/webapp/content/scss/themes/_dark-variables.scss:346]: Please use default colors here. We have themed presets for "success" (answer correct) and "danger" (answer wrong). The add container colors seems to be some sort of blue, please use the "primary" color here.
If you need to use custom colors, please prepare a very good reason and justify this, as documented at https://docs.artemis.cit.tum.de/dev/guidelines/client-design/

File: src/main/webapp/app/exercises/quiz/shared/questions/multiple-choice-question/multiple-choice-question.component.scss [hunk 3]
```diff
@@ -96,3 +96,111 @@
...
+.visual-answer-add-container-disabled {
+    background: var(--quiz-visual-answer-add-container-hover-color);
+    align-items: center;
+    display: flex;
+    justify-content: center;
+    cursor: pointer;
+    width: 2.5%;
+    margin-left: 0.2rem;
+    border-radius: 2px;
+    color: white;
```
Comment (bassner) [inline src/main/webapp/app/exercises/quiz/shared/questions/multiple-choice-question/multiple-choice-question.component.scss:185]: Please do not use hardcoded colors

File: src/test/javascript/spec/component/quiz-exercise/quiz-exercise-visual-editor.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    });
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    it('should initialize', fakeAsync(() => {
+        fixture.detectChanges();
+        expect(comp).not.toBeNull();
+    }));
```
Comment (bassner) [inline src/test/javascript/spec/component/quiz-exercise/quiz-exercise-visual-editor.component.spec.ts:54]: This test is useless. I think we can remove it. All other tests will fail if the component can't be created

File: src/test/javascript/spec/component/quiz-exercise/quiz-exercise-visual-editor.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    it('parse the given question properly to markdown', fakeAsync(() => {
+        fixture.detectChanges();
+        expect(comp).not.toBeNull();
```
Comment (bassner) [inline src/test/javascript/spec/component/quiz-exercise/quiz-exercise-visual-editor.component.spec.ts:58]: This is not needed, same in the other test. The beforeEach above will already fail if comp is really null

**Review by bassner — CHANGES_REQUESTED** [review]
Hey, please see the code comments below, thanks :)

**Review by rriyaldhi — DISMISSED** [review]
Functionality looks good. Tested on TS2.

**Review by danielbode — APPROVED** [review]
Tested on ts2. Reapproved.

**Review by bassner — CHANGES_REQUESTED** [review]
Hey!
Thank you for updating the code. I've tested this and have a few concerns:

1. The buttons don't look good on small screen widths (could happen on mobile phones):
image

2. The dark mode appearance of the editor is very different and doesn't look very polished. Why do we have a background color that separates it from the underlying background? In the white mode it's all white. This also leads to spacing issues; the text and the icons are directly pushed towards the border of that gray box. Also, there seems to not be a visible border box on the individual answer options like in the light mode. Please adapt the appearance to look similar in both themes; just the used colors should be inverted, basically.

image

3. I think using the blue "+" button on the right side of the question text input box of the new answer option is not a good approach. This can be confusing. What would a user think who sees this for the first time? What does the button do? Do I add another "question text" as it's next to the question text input field? What if I forget to click that little plus button? My new answer option won't be saved, right?

I recommend the same approach that is used to add new questions.
- Do not show the "New answer option" box at all. Instead, add a big green "+ Add answer option" below existing answer options (styled like the "Add Multiple-Choice question" button).
- If clicked, add a new empty answer option. It's actually added to the question, but simply empty at the beginning. The user can then edit it. Everything that is typed there will be saved when "Save" is clicked, without the need to click another button just to confirm that new answer option.

**Comment by sebastiannikles** [comment]
Should all be updated as requested. The problem with the background color was a general thing with the parent component. Should work now.

File: src/main/webapp/content/scss/themes/_dark-variables.scss (hunk 2 above)
Comment (bassner) [inline src/main/webapp/content/scss/themes/_dark-variables.scss:346]: One last thing, sorry.
Seems like you updated the definitions to default colors in the light theme only in 5ccb99481c94f136b5cf9529d4eace2429f46dfe. We need to do this for the dark theme, too.

**Review by bassner — CHANGES_REQUESTED** [review]

**Review by bassner — APPROVED** [review]
LGTM!

**Pull request merged**